### PR TITLE
Avoid duplicated boot of test kernel

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes -
| Related issues/PRs | part of #4798 
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Avoid duplicate boot of test kernel.

#### Why?

In Symfony its not longer allowed that TestKernel is booted multiple times. Currently in most cases our TestKernel is booted again in `createClient` so we need to move that at the begin in most setups.

#### To Do

- [ ] ~~Create a documentation PR~~
- [ ] ~~Add breaking changes to UPGRADE.md~~

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php [hunk 1]
```diff
@@ -53,7 +59,8 @@ public function testGetConfig()
 
     public function testGetNotExistingMetdata()
     {
-        $client = $this->createAuthenticatedClient();
+        $client = $this->client;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Tests/Functional/Controller/AdminControllerTest.php:62]: Instead of writing it like this I would prefer to use `$this->client` everywhere. But if it is too much work to change everywhere this works for me as well.

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/TargetGroupControllerTest.php [hunk 2]
```diff
@@ -374,15 +379,17 @@ public function testSingleDelete()
 
         $this->getEntityManager()->flush();
 
-        $client = $this->createAuthenticatedClient();
-        $client->request('DELETE', self::BASE_URL . '/' . $targetGroup->getId());
+        $targetGroupId = $targetGroup->getId();
```
Comment (danrot) [inline src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/TargetGroupControllerTest.php:382]: Is there any specific reason to extract that into a separate variable? In tests I usually try to avoid something like this, because if the `id` of the target group changes for some reason the testing logic might be flawed.

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php [hunk 3]
```diff
@@ -1436,26 +1439,28 @@ public function testDelete()
...
         $client->request(
             'DELETE',
-            '/api/categories/' . $category->getId()
+            '/api/categories/' . $categoryId
         );
 
         $this->assertHttpStatusCode(204, $client->getResponse());
 
-        $client = $this->createAuthenticatedClient();
+        $client = $this->client;
```
Comment (danrot) [inline src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php:1452]: This line would not be necessary, would it?

File: src/Sulu/Bundle/ContactBundle/Tests/Functional/Controller/ContactControllerTest.php [hunk 4]
```diff
@@ -1877,20 +1873,22 @@ public function testDelete()
...
+        $contactId = $contact->getId();
+
+        $client = $this->client;
+        $client->request('DELETE', '/api/contacts/' . $contactId);
 
         $this->assertHttpStatusCode(204, $client->getResponse());
 
-        $client = $this->createTestClient();
-        $client->request('GET', '/api/contacts/' . $contact->getId());
+        $client = $this->client;
```
Comment (danrot) [inline src/Sulu/Bundle/ContactBundle/Tests/Functional/Controller/ContactControllerTest.php:1883]: Would also not be necessary. If we would always be using `$this->client` this error also would not happen :see_no_evil:

File: src/Sulu/Bundle/CoreBundle/Tests/Functional/Controller/LocalizationControllerTest.php [hunk 5]
```diff
@@ -11,25 +11,14 @@
...
-    public function setUp(): void
-    {
-        $this->em = $this->getEntityManager();
-        $this->purgeDatabase();
-    }
-
     public function testCgetAction()
     {
         $client = $this->createAuthenticatedClient();
+        $this->purgeDatabase();
```
Comment (danrot) [inline src/Sulu/Bundle/CoreBundle/Tests/Functional/Controller/LocalizationControllerTest.php:21]: Shouldn't we keep that in the `setUp`, in case there'll be more tests added?

File: src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php [hunk 6]
```diff
@@ -85,12 +85,20 @@ public function onPostSerialize(ObjectEvent $event)
             $customUrlProperty
         );
 
-        $creatorFullName = $this->userManager->getFullNameByUserId($customUrl->getCreator());
+        $creatorFullName = null;
+        if ($creator = $customUrl->getCreator()) {
```
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php:89]: This is already part of #5242, right?

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php [hunk 7]
```diff
@@ -113,7 +113,13 @@ function(BundleInterface $bundle) {
                 $this->kernel->getBundles()
             );
 
-            $directories[] = $this->kernel->getRootDir();
+            if (method_exists($this->kernel, 'getRootDir')) {
+                $directories[] = $this->kernel->getRootDir();
+            }
+
+            $directories[] = $this->kernel->getProjectDir();
+            $directories[] = $this->kernel->getProjectDir() . '/src';
```
Comment (danrot) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php:121]: This is also part of another PR, isn't it?

File: src/Sulu/Bundle/MediaBundle/Tests/Application/Kernel.php [hunk 8]
```diff
@@ -20,6 +21,10 @@ public function registerContainerConfiguration(LoaderInterface $loader)
     {
         parent::registerContainerConfiguration($loader);
 
-        $loader->load(__DIR__ . '/config/config.yml');
+        if (SuluKernel::CONTEXT_WEBSITE === $this->getContext()) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Application/Kernel.php:24]: Why was that necessary?

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php [hunk 9]
```diff
@@ -341,11 +347,10 @@ protected function setUpCollection()
         $this->collection->setStyle(json_encode($style));
 
         // Create Collection Type
-        $this->collectionType = new CollectionType();
-        $this->collectionType->setName('Default Collection Type');
-        $this->collectionType->setDescription('Default Collection Type');
+        $collectionTypeFixtures = new LoadCollectionTypes();
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php:350]: How many collection types does this create? What if just one would be enough for this test? Then this test is much slower now, isn't it?

File: src/Sulu/Bundle/PageBundle/Tests/Functional/Controller/PageControllerTest.php [hunk 10]
```diff
@@ -42,7 +43,6 @@ class PageControllerTest extends SuluTestCase
 
     public function setUp(): void
     {
-        parent::setUp();
```
Comment (danrot) [inline src/Sulu/Bundle/PageBundle/Tests/Functional/Controller/PageControllerTest.php:45]: Aren't we missing something because of that?

File: src/Sulu/Bundle/RouteBundle/Tests/Unit/Generator/SymfonyExpressionTokenProviderTest.php [hunk 11]
```diff
@@ -55,6 +55,8 @@ public function testResolveNotExists()
     {
         $this->expectException(CannotEvaluateTokenException::class);
         $translator = $this->prophesize(Translator::class);
+        $translator->getLocale()->willReturn('en');
+        $translator->setLocale('en')->shouldBeCalled();
```
Comment (danrot) [inline src/Sulu/Bundle/RouteBundle/Tests/Unit/Generator/SymfonyExpressionTokenProviderTest.php:59]: Why did that change?

File: src/Sulu/Bundle/SearchBundle/Tests/Functional/Controller/SearchControllerTest.php [hunk 12]
```diff
@@ -43,9 +43,8 @@ class SearchControllerTest extends SuluTestCase
 
     public function setUp(): void
     {
-        parent::setUp();
```
Comment (danrot) [inline src/Sulu/Bundle/SearchBundle/Tests/Functional/Controller/SearchControllerTest.php:46]: Can we just remove that?

File: src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php [hunk 13]
```diff
@@ -55,10 +55,9 @@ class SnippetControllerTest extends SuluTestCase
 
     public function setUp(): void
     {
-        parent::setUp();
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php:58]: Can we just remove that?

File: src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php [hunk 14]
```diff
@@ -78,7 +79,7 @@ public function clear()
 
             $this->cacheManager->invalidateDomain($request->getHost());
 
-            $this->eventDispatcher->dispatch(Events::CACHE_CLEAR);
+            $this->eventDispatcher->dispatch(new CacheClearEvent(), Events::CACHE_CLEAR);
```
Comment (danrot) [inline src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php:82]: I thought they changed this call to only have on parameter? :thinking:

Comment (danrot) [reply src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php:82]: And how is this related to this PR anyway?

**Comment by danrot** [comment]
Oh, and I am not sure if you've realized, but the tests are not working.

**Comment by alexander-schranz** [comment]
> 1. I would avoid lines like $client = $this->client. I can imagine that you've done that because it is faster, but would removing them not be executing two regex?

Also because of git history and the code review it would make it really hard else to see what I also needed to change.

> 2. You have extracted quite some IDs from entities in the tests? Why is that? I tried to avoid storing values like that in tests, because it might hide some bugs (imagine the value is changes by the system under test in the mean time).

That it basically needed now as the Test itself share the same Kernel as the Request they share the same Doctrine Unit of Work and if you remove a entity from the database your stored `$entity` variable will not longer have its id `$entity->getId()` is null until its removed. As in most deletions we do after also a `GET` request on e.g. `/api/media/ . $media->getId()` will now return `/api/media/` because getId is null. So that we can really do that check we need to store that `id` in a seperate variable.

File: src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/TargetGroupControllerTest.php (hunk 2 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AudienceTargetingBundle/Tests/Functional/Controller/TargetGroupControllerTest.php:382]: commented below. basically `$targetGroup->getId();` will return null after it was removed because they share now the same doctrine unit of work.

File: src/Sulu/Bundle/MediaBundle/Tests/Application/Kernel.php (hunk 8 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Tests/Application/Kernel.php:24]: To run MediaStreamControllerTest and MediaWebsiteControllerTest under the correct WebsiteKernel as they test website routes and not admin routes.

File: src/Sulu/Bundle/PageBundle/Tests/Functional/Controller/PageControllerTest.php (hunk 10 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/PageBundle/Tests/Functional/Controller/PageControllerTest.php:45]: No there is no parent implimentation and as most TestCases don't call it I removed it also from this ones to be consistent.

File: src/Sulu/Bundle/RouteBundle/Tests/Unit/Generator/SymfonyExpressionTokenProviderTest.php (hunk 11 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/RouteBundle/Tests/Unit/Generator/SymfonyExpressionTokenProviderTest.php:59]: The translator does not longer accept `null` as locale and did fail. Related to this changes #5137

File: src/Sulu/Bundle/SearchBundle/Tests/Functional/Controller/SearchControllerTest.php (hunk 12 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SearchBundle/Tests/Functional/Controller/SearchControllerTest.php:46]: yes they is no parent implementation I removed it to be consistent

File: src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php (hunk 13 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SnippetBundle/Tests/Functional/Controller/SnippetControllerTest.php:58]: yes they is no parent implementation I removed it to be consistent

File: src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php (hunk 14 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php:82]: Oh sure I wanted to move this into an own PR 🙈 . 
No it are still 2 parameters the second one the string is optional and needed to be keeped for backward compatibility.

**Comment by danrot** [comment]
1. I think I would not make code "weird" because we don't want to change the git history. And GitHub would probably do a good job on higlighting all the added `$this->` parts, so I think I would be able to handle the review.
2. That totally makes sense :+1:

File: src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php (hunk 14 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/WebsiteBundle/Cache/CacheClearer.php]: Own PR created #5247

File: src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php (hunk 9 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Tests/Functional/Controller/MediaControllerTest.php:350]: It create the 2 types (Default and System) some test failed now because the system collection was not there or had a false id 🤔

File: src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php (hunk 7 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/DocumentManagerBundle/Command/FixturesLoadCommand.php]: Removed from this PR

File: src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php (hunk 6 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php]: Removed from this PR

File: src/Sulu/Bundle/CoreBundle/Tests/Functional/Controller/LocalizationControllerTest.php (hunk 5 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/CoreBundle/Tests/Functional/Controller/LocalizationControllerTest.php:27]: Did move it to the setup.

File: src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php (hunk 3 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/CategoryBundle/Tests/Functional/Controller/CategoryControllerTest.php]: Removed all appearance of `$client = $this->client;`

**Comment by danrot** [comment]
Still wondering if the file "src/Sulu/Bundle/MediaBundle/Tests/Unit/Media/Manager/test.txt" is necessary...

**Comment by danrot** [comment]
And some tests are still failing.

**Comment by alexander-schranz** [comment]
@danrot The `test.txt` is needed because we need to construct the UploadedFile for the test and Symfony will do in Symfony 5 a file_exists check. 

Fixed the tests but still would need #5242 that the CustomUrlBundle tests work.

**Comment by alexander-schranz** [comment]
@danrot lowest tests are also fixed.

**Comment by danrot** [comment]
Would have been nice if you would have done that in a separate commit :confused: But from what I've seen the fix seems to be some `clear` calls on the entity manager, is that right? Why was that necessary?

**Comment by alexander-schranz** [comment]
@danrot its a seperate commit: https://github.com/sulu/sulu/pull/5244/commits/bec2236eeab274f098eeaf677053dea4d9599819

**Comment by alexander-schranz** [comment]
The unit of work seems to bettoo buggy in the lower versions to be shared between the and think its also good to clear the entity manager else you can not be sure if its really loaded from database and not from the unit of work when requested a edit item.
